test/index.ts

```diff
@@ -557,23 +557,55 @@ function testNonEmptyArray() {
 }
 
 function testReadonlyKeys() {
-  type T = { readonly a: number; b: string };
-
-  type Actual = ReadonlyKeys<T>;
-
-  type Expected = "a";
-
-  type Test = Assert<IsExact<Actual, Expected>>;
+  type cases = [
+    // @ts-expect-error primitives not allowed
+    Assert<IsExact<ReadonlyKeys<string>, keyof string>>,
+    // @ts-expect-error primitives not allowed
+    Assert<IsExact<ReadonlyKeys<number>, keyof number>>,
+    // @ts-expect-error primitives not allowed
+    Assert<IsExact<ReadonlyKeys<boolean>, keyof boolean>>,
+    // @ts-expect-error primitives not allowed
+    Assert<IsExact<ReadonlyKeys<bigint>, keyof bigint>>,
+    // @ts-expect-error primitives not allowed
+    Assert<IsExact<ReadonlyKeys<symbol>, keyof symbol>>,
+    // @ts-expect-error primitives not allowed
+    Assert<IsExact<ReadonlyKeys<null>, keyof null>>,
+    // @ts-expect-error primitives not allowed
+    Assert<IsExact<ReadonlyKeys<undefined>, keyof undefined>>,
```

Comment:
you don't need `Assert` for this test cases since then will throw a TypeError, therefore I suggest removing `Assert` and expected values:

```suggestion
    // @ts-expect-error primitives not allowed
   ReadonlyKeys<string>,
    // @ts-expect-error primitives not allowed
    ReadonlyKeys<number>,
    // @ts-expect-error primitives not allowed
    ReadonlyKeys<boolean>,
    // @ts-expect-error primitives not allowed
    ReadonlyKeys<bigint>,
    // @ts-expect-error primitives not allowed
    ReadonlyKeys<symbol>,
    // @ts-expect-error primitives not allowed
    ReadonlyKeys<null>,
    // @ts-expect-error primitives not allowed
    ReadonlyKeys<undefined>,
```

Reply:
Yup, apologies I missed that!

@Beraliv Also, we currently have the `object` constraint in `ReadonlyKeys` and `WritableKeys`, but not in `OptionalKeys` and `RequiredKeys`. I feel we should add the constraint to the latter two types as well. WDYT?

Reply:
Yeah, I think it should be aligned, but at the same time it introduces a breaking change. For the time being, please don't align it. Next time, we're making a major release, I'm happy to update it

Reply:
Cool!!

Also, there's scope for adding constraints to type parameters in several other types as well, like the `Mark*` family of types. It may already be in the plan, but if not, we should probably think about this during the next major release.

Reply:
I'll create an issue where we can discuss all major changes. For now, I'm going to release patch changes only